feat: have all config/consts together, and print on startup

franciscodiazydiaz opened the pull request:
## 📝 Summary

Address issue #181

## ⛱ Motivation and Context

* All config/consts should be together
* In particular which can be set through env vars
* The config should be printed on startup
* Keep changes backward compatible

## 📚 References

* [spf13/viper](https://github.com/spf13/viper)
* [Same flag name for different cobra commands does not work](https://github.com/spf13/viper/issues/233)

## TODO

* ~[ ] Define the output format of the config to be printed during startup~
* [x] Write tests to validate the behavior
* [ ] Website & API share the same `LISTEN_ADDR` environment variable
* ~[ ] Sensitive data should be masked or skipped when printing the configuration~
* [x] Update README

---

## ✅ I have run these commands

* [x] `make lint`
* [x] `make test-race`
* [x] `go mod tidy`
* [x] I have seen and agree to `CONTRIBUTING.md`

metachris commented:
This is amazing!

I'll take time to review in depth tomorrow.

One thought -- should we use consts for the names, rather than reuse strings (that could easily have a spelling issue)? 

Like https://github.com/flashbots/mev-boost-relay/pull/205/files#diff-16898761923b521dcddba86f2ee719287bcf460344f2c858b976bac13c7d0be9R57

```go
	if config.GetString("allowSyncingBeaconNode") != "" {
```

Alos, let's keep the PRs small if possible -- i.e. don't add the printing of the config here, but in another pr.

franciscodiazydiaz commented:
You are right, indeed if you misspell the name it will behave as if the variable is empty, and it won't fail.

I'll remove the code that prints the config.

franciscodiazydiaz commented:
Just to clarify, is something like this what you have in mind? Probably I'm overthinking :P 

```go
// config/variables.go
const (
 AllowSyncingBeaconNode = "AllowSyncingBeaconNode"
...
)

// beaconclient/multi_beacon_client.go
if config.GetString(config.AllowSyncingBeaconNode) != "" {
```

metachris commented:
Yes, had exactly that in mind 👍

metachris commented:
let's get this over the line @franciscodiazydiaz ? 🙏 🚀

franciscodiazydiaz commented:
I went through the code several times and hopefully, I haven't missed anything.

Changing the variable `LISTEN_ADDR` in Website & API will introduce a backward incompatibility, that could be tackled in another PR.

metachris commented on README.md:
```diff
@@ -100,17 +100,56 @@ redis-cli DEL boost-relay/sepolia:validators-registration boost-relay/sepolia:va
+| RELAY_URL | full url for the relay (https://pubkey@host) | `""` |
+| PUBKEY_OVERRIDE | override for public key | `""` |
+
+#### Database
+
+| Name | Description | Default |
+|------|-------------|---------|
+| DB_TABLE_PREFIX | prefix to use for db tables | `dev` |
+| DB_DONT_APPLY_SCHEMA | disable applying DB schema on startup (useful for connecting data API to read-only replica) | `""` |
+| PRINT_SCHEMA | print DB schema | `false` |
```
This should be part of the common section

metachris commented on config/variables.go:
```diff
@@ -0,0 +1,172 @@
+	bindAndSet(AllowSyncingBeaconNode, "ALLOW_SYNCING_BEACON_NODE", "")
+}
+
+func bindAndSet(key, envVariable string, defaultValue interface{}) {
```
```suggestion
func bindAndSet(key, envVariable string, defaultValue any) {
```

franciscodiazydiaz replied:
Didn't know `any` 👍

metachris commented on config/variables.go:
```diff
@@ -0,0 +1,172 @@
+}
+
+// Get returns an interface. For a specific value use one of the Get____ methods.
+func Get(key string) interface{} { return viper.Get(key) }
```
is this needed/used?

franciscodiazydiaz replied:
Not really. I'll remove it.

metachris commented on config/variables.go:
```diff
@@ -0,0 +1,172 @@
+	WebsiteDefaultPubkeyOverride    = ""
+
+	// Common: Api, HouseKeeper & Website
+	Network        = "Network"
```
I wonder if we should prefix them with `Key` -- like `KeyNetwork` to make it clear it's the key, and not the network, when using it.

franciscodiazydiaz replied:
Sure, I can give it a try

metachris commented:
Will test this in testnets and merge end of the week.

Please rebase with main (now have a conflict in go.sum) 🙏

franciscodiazydiaz commented:
Cool! I'll be around if you need something from my side.

metachris commented:
i wanted to resolve a merge confict, and GitHub added a commit to your branch, but i think that's not doing all that's needed (still need to use the new postgres setup in `tool export-builder-bids`).

But more importantly, when trying to run this branch, i get this error:

```panic: data-api-export-payloads flag redefined: db
goroutine 1 [running]:
github.com/spf13/pflag.(*FlagSet).AddFlag(0xc00012e500, 0xc000193900)
        github.com/spf13/pflag@v1.0.5/flag.go:848 +0x5e6
github.com/spf13/pflag.(*FlagSet).VarPF(0x10e54a0?, {0xcc6ca8, 0xc00016ba90}, {0xba8279, 0x2}, {0x0, 0x0}, {0xbae7d8, 0xe})
        github.com/spf13/pflag@v1.0.5/flag.go:831 +0x151
github.com/spf13/pflag.(*FlagSet).VarP(...)
        github.com/spf13/pflag@v1.0.5/flag.go:837
github.com/spf13/pflag.(*FlagSet).StringVarP(0xba8578?, 0x3?, {0xba8279?, 0x0?}, {0x0?, 0xbaf549?}, {0x0?, 0xba8a0a?}, {0xbae7d8, 0xe})
        github.com/spf13/pflag@v1.0.5/string.go:42 +0x6e
github.com/spf13/pflag.(*FlagSet).String(...)
        github.com/spf13/pflag@v1.0.5/string.go:60
github.com/flashbots/mev-boost-relay/cmd/tool.init.2()
        github.com/flashbots/mev-boost-relay/cmd/tool/export-data-api-payloads-delivered.go:18 +0x6e```

franciscodiazydiaz commented:
Don't worry, I'll push the fix today. The command `export-data-api-payloads-bids` uses the flag definition from `export-data-api-payloads-delivered` (DataAPIExportPayloads) causing the redefinition of db.

metachris commented:
I apologize that merging the other PR (#224) again caused a bunch of conflicts!

This PR is the next one to be merged. Would very much appreciate if you find the time for a final rebase 🙏 Sorry for the inconvenience, I know this has been in the making for a long time. I think we can merge it today. Thanks for your efforts! <3

franciscodiazydiaz commented:
It is fine; I don't mind fixing these issues as it lets me know more about the code. I'll have it ready for today.

metachris commented:
@franciscodiazydiaz would you reach out to me on the Flashbots Discord or [TG](https://t.me/metachris)?

metachris commented:
Asked @0xpanoramix to review this PR as well since he's got extensive experience with Viper and Cobra (way more than me ^^). Shouldn't take long, but in the meantime i've merged another PR that caused a minor conflict 🙏

0xpanoramix commented on config/variables.go:
```diff
@@ -0,0 +1,171 @@
+	KeyDBTestDSN         = "KeyKeyDBTestDSN"
+
+	// Datastore
+	KeyDisableBidMemoryCache = "KeyDisableBidMemoryCache"
```
`KeyDisableBidMemoryCache` is unused right now, I suppose it should be ?

franciscodiazydiaz replied:
Good catch!

franciscodiazydiaz replied:
The variable was removed in https://github.com/flashbots/mev-boost-relay/pull/206. I miss this during some rebase.

0xpanoramix commented on config/variables.go:
```diff
@@ -0,0 +1,171 @@
+	for k, v := range configEnvs {
+		config[v] = viper.GetString(k)
+	}
+	// FIXME: Needs to mask or skip sensitive data
```
Is this comment still relevant ?

franciscodiazydiaz replied:
Right now the function is not being used. I'll remove it and leave it for a different PR.

0xpanoramix reviewed (commented):
If the architecture looks good to you @metachris, the only general comment I'd have is to add a quick documentation in the code explaining how to add new commands and flags @franciscodiazydiaz.
Otherwise, LGTM !

franciscodiazydiaz commented:
@0xpanoramix, thanks for the feedback! I've removed the unused code and added some documentation to the config, let me know if that is what you have in mind or if it is not clear enough.

metachris commented:
I'm afraid that I can't comprehensively test this PR and that I might miss any flags / env vars :(

franciscodiazydiaz commented:
No worries, I know this PR is touching many places that could be complex to test. I can close the PR, and if I can come up with a way to comprehensively test it I will open it again.
